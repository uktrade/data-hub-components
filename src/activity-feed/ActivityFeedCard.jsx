import React from 'react'
import { find } from 'lodash'

import PropTypes from 'prop-types'
import Cards from './cards'

export default class ActivityFeedCard extends React.PureComponent {
  static propTypes = {
    activity: PropTypes.object.isRequired,
    showDetails: PropTypes.bool,
  }

  static defaultProps = {
    showDetails: false,
  }

  constructor(props) {
    super(props)
    const { activity } = this.props
    this.Card = find(Cards, c => c.canRender(activity))
  }

  render() {
    const { activity, showDetails } = this.props
    const { Card } = this

    return Card ? <Card activity={activity} showDetails={showDetails} /> : null
  }
}
