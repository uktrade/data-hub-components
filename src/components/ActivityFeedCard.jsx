import React from 'react'
import PropTypes from 'prop-types'
import { Link, InsetText, Table, Paragraph, H3 } from 'govuk-react'
import { filter, includes, get, map } from 'lodash'
import moment from 'moment'

const getPeople = (activity, personSubType) => {
  return map(filter(activity['object']['attributedTo'], ({ type }) => {
      return includes(type, `dit:${personSubType}`)
    }), ({ id, name, 'dit:emailAddress': emailAddress }) => {
      return {
        id,
        name,
        emailAddress,
      }
    })
}

export default class ActivityFeedCard extends React.Component {
  static propTypes = {
    activity: PropTypes.object,
  }

  constructor(props) {
    super(props)

    this.state = {
      showMoreContent: false,
    }

    this.onMoreContentClick = this.onMoreContentClick.bind(this)
  }

  onMoreContentClick(event) {
    event.preventDefault()

    this.setState({ showMoreContent: !this.state.showMoreContent })
  }

  render() {
    const { activity } = this.props
    const published = moment(activity.published).fromNow()
    const advisers = getPeople(activity, 'Adviser')
    const subject = get(activity, 'object.dit:subject')
    const service = get(activity, 'object.dit:service.name')
    const url = get(activity, 'object.url')

    return (
      <div style={{ border: '1px solid #c0c0c0', padding: '10px' }}>
        <div style={{ display: 'flex', flexFlow: 'row wrap' }}>
          <div style={{ flexGrow: 1, marginBottom: '10px' }}>
            <Paragraph>An interaction with the company took place</Paragraph>
          </div>
          <div style={{ flexGrow: 1, marginBottom: '10px', textAlign: 'right' }}>
            <Paragraph>{published}</Paragraph>
          </div>
          <div style={{ width: '100%', marginBottom: '10px' }}>
            <H3>{subject}</H3>
          </div>
        </div>
        <Link href="#" onClick={this.onMoreContentClick}>Who was involved</Link>
        {
          this.state.showMoreContent ? <div>
            <InsetText>
              <Table>
                <Table.Row>
                  <Table.CellHeader>Advisers</Table.CellHeader>
                  <Table.Cell>
                    {advisers.map(({ id, name, emailAddress }) => {
                      return <React.Fragment key={id}>
                          <span>
                            {name} <Link href={"mailto:" + emailAddress}>{emailAddress}</Link>
                          </span>
                      </React.Fragment>
                    })}
                  </Table.Cell>
                </Table.Row>
                <Table.Row>
                  <Table.CellHeader>DIT team</Table.CellHeader>
                  <Table.Cell>
                    team
                  </Table.Cell>
                </Table.Row>
                <Table.Row>
                  <Table.CellHeader>Services</Table.CellHeader>
                  <Table.Cell>{service}</Table.Cell>
                </Table.Row>
              </Table>
              <Link href={url}>Go to the interaction detail page</Link>
            </InsetText>
          </div> : null
        }
      </div>
    )
  }
}
