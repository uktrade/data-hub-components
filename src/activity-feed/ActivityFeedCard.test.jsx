import React from 'react'
import renderer from 'react-test-renderer'
import moment from 'moment'
import { set } from 'lodash'

import ActivityFeedCard from './ActivityFeedCard'
import interactionActivityFixture from '../../fixtures/activity_feed/interactions/interaction'
import serviceDeliveryActivityFixture from '../../fixtures/activity_feed/interactions/service_delivery'
import investmentProjectsCTIFixture from '../../fixtures/activity_feed/investment_projects/project_added_cti'
import investmentProjectsFDIFixture from '../../fixtures/activity_feed/investment_projects/project_added_fdi'
import investmentProjectsNonFDIFixture from '../../fixtures/activity_feed/investment_projects/project_added_non_fdi'
import orderAddedFixture from '../../fixtures/activity_feed/omis/order_added'

import MockDate from 'mockdate'

// Lock the date so moment's relative date doesn't break our deterministic tests.
MockDate.set(1559750582706)

describe('ActivityFeedCard', () => {
  describe('when the interaction is empty', () => {
    test('should render null', () => {
      const tree = renderer
        .create(<ActivityFeedCard activity={{}}/>)
        .toJSON()
      expect(tree).toBeNull()
    })
  })

  describe('when there is an interaction', () => {
    test('should render interaction activity', () => {
      const tree = renderer
        .create(<ActivityFeedCard activity={interactionActivityFixture} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when the interaction is a draft in the past', () => {
    test('should render interaction activity with "Incomplete interaction" badge', () => {
      const fixture = { ...interactionActivityFixture };
      set(fixture, 'object.dit:status', 'draft')
      set(fixture, 'object.startTime', moment().subtract(1, 'years').toISOString())
      const tree = renderer
        .create(<ActivityFeedCard activity={fixture} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when the interaction is a draft and upcoming', () => {
    test('should render interaction activity with "Upcoming interaction" badge', () => {
      const fixture = { ...interactionActivityFixture };
      set(fixture, 'object.dit:status', 'draft')
      set(fixture, 'object.startTime', moment().add(1, 'days').toISOString())
      const tree = renderer
        .create(<ActivityFeedCard activity={fixture} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when the interaction is a draft and archived', () => {
    test('should render interaction activity with "Cancelled interaction" badge', () => {
      const fixture = { ...interactionActivityFixture };
      set(fixture, 'object.dit:status', 'draft')
      set(fixture, 'object.dit:archived', true)
      const tree = renderer
        .create(<ActivityFeedCard activity={fixture} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when the interaction is complete', () => {
    test('should render interaction activity with "Completed interaction" badge', () => {
      const fixture = { ...interactionActivityFixture };
      set(fixture, 'object.dit:status', 'complete')
      const tree = renderer
        .create(<ActivityFeedCard activity={fixture} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when there is a service delivery', () => {
    test('should render service delivery activity', () => {
      const tree = renderer
        .create(<ActivityFeedCard activity={serviceDeliveryActivityFixture} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when there is an activity item of arbitrary type', () => {
    test('should render default activity', () => {
      const tree = renderer
        .create(<ActivityFeedCard activity={{
          object: {
            'dit:subject': 'subject',
          }
        }}/>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when there is an interaction without any people involved', () => {
    test('should render interaction activity without advisers nor contacts', () => {
      const interactionWithNoPeople = { ...interactionActivityFixture }
      set(interactionWithNoPeople,
        'object.attributedTo',
        interactionActivityFixture.object.attributedTo.filter(a => !a.type.includes('Person'))
      )

      const tree = renderer
        .create(<ActivityFeedCard activity={interactionWithNoPeople} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when the interaction does not have a service', () => {
    test('should render interaction activity without services', () => {
      const interactionWithoutService = { ...interactionActivityFixture }
      set(interactionWithoutService, 'object.dit:service', null)

      const tree = renderer
        .create(<ActivityFeedCard activity={interactionWithoutService} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when an investment project is CTI', () => {
    test('should render the CTI investment project activity card', () => {
      const commitmentToInvest = { ...investmentProjectsCTIFixture }
      const tree = renderer
        .create(<ActivityFeedCard activity={commitmentToInvest} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when an investment project is FDI', () => {
    test('should render the FDI investment project activity card', () => {
      const foreignDirectInvestment = { ...investmentProjectsFDIFixture }
      const tree = renderer
        .create(<ActivityFeedCard activity={foreignDirectInvestment} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when an investment project is Non-FDI', () => {
    test('should render the Non-FDI investment project activity card', () => {
      const nonForeignDirectInvestment = { ...investmentProjectsNonFDIFixture }
      const tree = renderer
        .create(<ActivityFeedCard activity={nonForeignDirectInvestment} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when a New OMIS order is added', () => {
    test('should render an activity card', () => {
      const orderAdded = { ...orderAddedFixture }
      const tree = renderer
        .create(<ActivityFeedCard activity={orderAdded} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})




