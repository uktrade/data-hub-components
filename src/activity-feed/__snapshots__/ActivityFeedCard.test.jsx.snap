// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityFeedCard when there is a service delivery should render service delivery activity 1`] = `
<div
  className="sc-cmthru jjutsD"
>
  <div
    className="sc-hMFtBS bSjygG"
  >
    <div
      className="sc-cLQEGU ciAtZf"
    >
      A service was provided to the company
    </div>
    <div
      className="sc-gqPbQI fPqDmG"
    >
      a month ago
    </div>
    <div
      className="sc-hORach cvIZEb"
    >
      <h3
        className="sc-cMljjf kCLKvB"
        size="MEDIUM"
      >
        Openening of overseas office
      </h3>
    </div>
  </div>
  <details
    className="sc-bMVAic cchcEg sc-kEYyzF ihcnje"
    open={false}
  >
    <summary
      className="sc-kkGfuU cCdZFh"
    >
      <span
        className="sc-iAyFgw jRkIxn"
      >
        Who was involved
      </span>
    </summary>
    <div
      className="sc-hSdWYo dfimje"
    >
      <table
        className="sc-bwCtUz iVExri"
      >
        <tbody
          className="sc-eXEjpC fuVqJC"
        >
          <tr
            className="sc-iQKALj ftJWrM"
          >
            <th
              className="sc-RefOD dxRfSY"
              style={
                Object {
                  "border": 0,
                  "fontWeight": "normal",
                }
              }
            >
              Advisers
            </th>
            <td
              className="sc-RefOD gzbQXK"
              style={
                Object {
                  "border": 0,
                }
              }
            >
              <span>
                Mario Bros
                 
                <a
                  className="sc-bwzfXH VXMCV"
                  href="mailto:mariobros@dit.com"
                  muted={false}
                >
                  mariobros@dit.com
                </a>
              </span>
            </td>
          </tr>
          <tr
            className="sc-iQKALj ftJWrM"
          >
            <th
              className="sc-RefOD dxRfSY"
              style={
                Object {
                  "border": 0,
                  "fontWeight": "normal",
                }
              }
            >
              Services
            </th>
            <td
              className="sc-RefOD gzbQXK"
              style={
                Object {
                  "border": 0,
                }
              }
            >
              Events - Overseas
            </td>
          </tr>
        </tbody>
      </table>
      <a
        className="sc-bwzfXH VXMCV"
        href="https://example.com/companies/a4ead59f-dfc8-e511-a5e6-e4115bead28a/interactions/854d8d31-7405-4625-a957-0ed2c3092caa"
        muted={false}
      >
        Go to the interaction detail page
      </a>
    </div>
  </details>
</div>
`;

exports[`ActivityFeedCard when there is an activity item of arbitrary type should render default activity 1`] = `
<div
  className="sc-bAeIUo hKMfQm"
>
  <div
    className="sc-iujRgT izOATC"
  >
    <h3
      className="sc-cMljjf kCLKvB"
      size="MEDIUM"
    >
      subject
    </h3>
  </div>
</div>
`;

exports[`ActivityFeedCard when there is an interaction should render interaction activity 1`] = `
<div
  className="sc-cmthru jjutsD"
>
  <div
    className="sc-hMFtBS bSjygG"
  >
    <div
      className="sc-cLQEGU ciAtZf"
    >
      An interaction with the company took place
    </div>
    <div
      className="sc-gqPbQI fPqDmG"
    >
      a month ago
    </div>
    <div
      className="sc-hORach cvIZEb"
    >
      <h3
        className="sc-cMljjf kCLKvB"
        size="MEDIUM"
      >
        Invited to Technology Roundtable
      </h3>
    </div>
  </div>
  <details
    className="sc-bMVAic cchcEg sc-kEYyzF ihcnje"
    open={false}
  >
    <summary
      className="sc-kkGfuU cCdZFh"
    >
      <span
        className="sc-iAyFgw jRkIxn"
      >
        Who was involved
      </span>
    </summary>
    <div
      className="sc-hSdWYo dfimje"
    >
      <table
        className="sc-bwCtUz iVExri"
      >
        <tbody
          className="sc-eXEjpC fuVqJC"
        >
          <tr
            className="sc-iQKALj ftJWrM"
          >
            <th
              className="sc-RefOD dxRfSY"
              style={
                Object {
                  "border": 0,
                  "fontWeight": "normal",
                }
              }
            >
              Advisers
            </th>
            <td
              className="sc-RefOD gzbQXK"
              style={
                Object {
                  "border": 0,
                }
              }
            >
              <span>
                Mario Bros
                 
                <a
                  className="sc-bwzfXH VXMCV"
                  href="mailto:mariobros@dit.com"
                  muted={false}
                >
                  mariobros@dit.com
                </a>
              </span>
            </td>
          </tr>
          <tr
            className="sc-iQKALj ftJWrM"
          >
            <th
              className="sc-RefOD dxRfSY"
              style={
                Object {
                  "border": 0,
                  "fontWeight": "normal",
                }
              }
            >
              Services
            </th>
            <td
              className="sc-RefOD gzbQXK"
              style={
                Object {
                  "border": 0,
                }
              }
            >
              Account Management
            </td>
          </tr>
        </tbody>
      </table>
      <a
        className="sc-bwzfXH VXMCV"
        href="https://example.com/companies/a4ead59f-dfc8-e511-a5e6-e4115bead28a/interactions/a066cc49-b679-4cbe-93c5-3ceba2f067ec"
        muted={false}
      >
        Go to the interaction detail page
      </a>
    </div>
  </details>
</div>
`;
